geometry/multipolygon-triangulate-gts.cpp

```diff
@@ -0,0 +1,197 @@
+ * *  Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * *  Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "dbglog/dbglog.hpp"
+#include "gtsutils.hpp"
+#include "utility/expect.hpp"
+#include "utility/small_map.hpp"
+
+#include "geometry/polygon.hpp"
+
+#include "multipolygon-triangulate-gts.hpp"
+
+namespace geometry
+{
+/// Helper struct for triangle extraction
+struct TriData
+{
+    const utility::small_map<GtsVertex*, ItPair>& vertexMap;
+    const math::MultiPolygon& mpolygon;
+    std::vector<TriangleItPair> triangles;
+};
+
+/// Helper to find vertex in vertexMap and throw error if not found
+inline ItPair
+    findVertexInMap(GtsVertex* v,
+                    const utility::small_map<GtsVertex*, ItPair>& vertexMap)
+{
+    auto it = vertexMap.find(v);
+    if (it == vertexMap.end())
+    {
+        LOGTHROW(err4, std::runtime_error)
+            << "Cannot find a vertex in added vertex map.";
+    }
+    return it->second;
+}
+
+/// Extracts triangle and adds it to `data.triangles` if center is in
+/// multipolygon
+gint extractTriangle(GtsTriangle* t, TriData* data)
+{
+    GtsVertex *v1, *v2, *v3;
+    gts_triangle_vertices(t, &v1, &v2, &v3);
+
+    // check that centroid is inside multipolygon
+    math::Point2 p1(v1->p.x, v1->p.y);
+    math::Point2 p2(v2->p.x, v2->p.y);
+    math::Point2 p3(v3->p.x, v3->p.y);
+    math::Point2d c = (p1 + p2 + p3) * (1.0 / 3);
+
+    if (geometry::insideMultiPolygon(data->mpolygon, c))
+    {
+        data->triangles.push_back({ findVertexInMap(v1, data->vertexMap),
+                                    findVertexInMap(v2, data->vertexMap),
+                                    findVertexInMap(v3, data->vertexMap) });
+    }
+    return {};
+}
+
+std::vector<TriangleItPair>
+    multipolygonTriangulateGts(const math::MultiPolygon& mpolygon)
+{
+    geometry::checkGtsInitialized();
+
+    GtsSurface* s = gts_surface_new(gts_surface_class(),
```

Comment:
Can we extract the constrained Delaunay triangulation to `delaunay.cpp` and use it here? It would be nice to have a general implementation. It used to be there but has been dropped some time ago.

Reply:
As suggested in PM, I introduced a function that outputs normal polygons:
https://github.com/melowntech/libgeometry/blob/d5a46b57597d56779049c8d41899f4b7e152a845/geometry/multipolygon-triangulate-gts.cpp#L197-L208

I agree, it would be nice to have a function for general CDT, but I would leave it for the developer that actually needs it. Moving the current function to "delaunay2d.cpp" would be a bit confusing, because the polygon triangulation is done in "triangulate.cpp". I kept it separately from "triangulate.cpp" to avoid mixing in gts code (which is separated in other parts of the library), but I am open to moving it anywhere you suggest.